### DISCUSSION TITLE
Large amounts of update or delete optimization ?

### DISCUSSION THREAD

**Discussion by khanakia** [ClickHouse/ClickHouse#62310] [Q&A]
I have a use case where we fetch lots of data from Google Adwords let say Campaigns and save to the campaigns table.

But I want resync the same data once a day so what I want is, if the record exists then update if not then insert.

**What is the best practice for this use case?**

- Method 1: Delete all the data for a user by **user_id** column and batch insert again all the data?
- Method 2:  Check the record if exists then send async update if not then send async insert?

Or is there any other method or best way so I can keep my clickhouse optimized while performing such operations ?

**Comment by khanakia** [comment]
@den-crane  do you have some suggestions?

**Reply by den-crane** [reply]
The only solution is ReplacingMergeTree, everything else is a garbage.

**Reply by khanakia** [reply]
As per docs `The engine differs from MergeTree in that it removes duplicate entries with the same sorting key value (ORDER BY table section, not PRIMARY KEY).`


But we need to deduplicate based on the primary field because in my case `id or key` field will always different even for the same duplicate record as Google Ads or Google Search Console API does not returns any ID field

This is my primary key
```
primary_key {
    columns = [column.task_id, column.dated, column.queried, column.url_id, column.country_iso3, column.device_id]
  }
 ```

**Comment by pengxiaoo** [comment]
in clickhouse, ReplacingMergeTree is the optimal solution for OLAP scenario which has periodical large amount upserts(update if key exists otherwise insert).

primary key in clickhouse is not like in OLTP databases. primary key is always a prefix of the columns specified in the ORDER BY clause. you can even omit primary key in your table definition. but, ORDER BY clause cannot be omitted.

the order of the columns specified in the ORDER BY clause matters. for best performance, they should be ordered by cardinality in ascending order. columns with high cardinality should not come first, which will bring bad data locality. you should double check whether your primary_key columns obeys this rule

primary_key {
    columns = [column.task_id, column.dated, column.queried, column.url_id, column.country_iso3, column.device_id]
  }

**Reply by khanakia** [reply]
No i am not having a performance problem.


## Problem
I am working on integrating Google Ads, Search Console API data into Clickhouse, but I face a challenge with keeping the data up-to-date due to frequent updates to historical data from Google. Originally, I considered deleting the data for the past month and re-syncing it; however, deletion queries are resource-intensive in Clickhouse.

Another approach I contemplated was partitioning the data. However, Clickhouse has a maximum partition limit of 50,000, and given the potential scale with thousands of users, this limit could be easily exceeded.

Using a ReplacingMergeTree table engine isn't feasible either, as Google doesn't always provide a unique ID or primary key for each record, which is necessary for this table engine.

Given these constraints, what are other viable options, or how should I approach solving this issue to ensure the data remains current and accurate without incurring excessive performance costs?

let say i sync the data for Oct 2024 and Now on Nov 2024 i will sync data again for Oct and Nov 2024 both.

So i want to deduplicate the Oct 2024 Data

Here is attached schema as you asked.
[schema.txt](https://github.com/user-attachments/files/18206737/schema.txt)


---------------------------------------
---------------------------------------

Regarding your suggestion about partition by month won't work too the problem with this approach is

i will have 10000 Users or more and each user can have 10 Tasks (aka Google Ads, Search Console Account Linked)

10000 * 10 = 100000 partition and i create partition by task_id so i will run out of max partition limit and parts too

**Reply by pengxiaoo** [reply]
given Google doesn't always provide a unique ID or primary key for each record. how do we determine if 2 records are duplicated? having exact the same (task_id, dated, queried, url_id, country_iso3, device_id) which is your primary key?

`The engine differs from MergeTree in that it removes duplicate entries with the same sorting key value (ORDER BY table section, not PRIMARY KEY).` 
this only means if primary key and sorting key are not the same(i.e. primary key is a subset and prefix of the whole sorting key), de-duplication are done by sorting key, which is a superset of primary key.

in your case, your primary key and sorting key are the same. so yes you can use ReplacingMergeTree, there is no reason you cannot use it. btw, if you don't want to compare and delete by yourself, then ReplacingMergeTree is the only option in clickhouse.

partition by (user_id, task_id) looks really weird. this will generate too many partitions. why cannot partition by month? how does a typical query looks like for this table? usually we design the table tailored for the query pattern

**Reply by khanakia** [reply]
Yes `(task_id, dated, queried, url_id, country_iso3, device_id)` this is the primary key

the problem with partitioning by month only is how will i delete the partition of a single user only and resync the data for that single user or task_id

If i delete the partition by month at the end of month it will delete all the data for all the users no ?

**Reply by inv2004** [reply]
I would suggest to start from understanding what queries you are going to run and perf of the queries. Because it is not a big problem to store data, but to have adequate perf on select is much harder task

**Reply by inv2004** [reply]
@khanakia
Why do you want to delete some records? ReplacingMT will replace them automatically

**Reply by khanakia** [reply]
I wanted to delete because 
- it does not guarantee rows will be deduplicated, and you should not rely on it. https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree#query-time-de-duplication--final
- Merging occurs in the background at an unknown time, so you can't plan for it. https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree
- when we select the records it shows duplicated data until deduplication is done by clickhouse in background which is not guaranteed

- Another question: Can we use eventTime to deduplicate ? https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree#replacingmergetree-parameters

**Comment by inv2004** [comment]
It is very common thing to have duplicates in CH. You have to build your
queries to avoid it

Regards,
